From 67a1f3faec8b1ee7bba7bb5a2858d46bb81e67bc Mon Sep 17 00:00:00 2001
From: Robert Kanter <rkanter@apache.org>
Date: Tue, 2 Feb 2016 15:19:35 -0800
Subject: [PATCH 1170/2748] MAPREDUCE-6620. Jobs that did not start are shown
 as starting in 1969 in the JHS web UI (haibochen
 via rkanter)

(cherry picked from commit ccbba4a4deb8ade54a04137c993526e461bcb46e)
(cherry picked from commit 358c49d830905e233dd61d9822d7339c854e6d79)

Change-Id: Ibf279fa2cad96f7efaac873f3f9169b5f0bb6c5c
---
 .../hadoop/mapreduce/v2/hs/webapp/HsJobBlock.java  |    2 +-
 .../hadoop/mapreduce/v2/hs/webapp/HsJobsBlock.java |    2 +-
 .../hadoop/mapreduce/v2/hs/webapp/dao/JobInfo.java |   12 +++++++++
 .../mapreduce/v2/hs/webapp/dao/TestJobInfo.java    |   26 +++++++++++++++++++-
 4 files changed, 39 insertions(+), 3 deletions(-)

diff --git a/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-hs/src/main/java/org/apache/hadoop/mapreduce/v2/hs/webapp/HsJobBlock.java b/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-hs/src/main/java/org/apache/hadoop/mapreduce/v2/hs/webapp/HsJobBlock.java
index 4a13e0b..80b50cb 100644
--- a/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-hs/src/main/java/org/apache/hadoop/mapreduce/v2/hs/webapp/HsJobBlock.java
+++ b/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-hs/src/main/java/org/apache/hadoop/mapreduce/v2/hs/webapp/HsJobBlock.java
@@ -85,7 +85,7 @@
         _("State:", job.getState()).
         _("Uberized:", job.isUber()).
         _("Submitted:", new Date(job.getSubmitTime())).
-        _("Started:", new Date(job.getStartTime())).
+        _("Started:", job.getStartTimeStr()).
         _("Finished:", new Date(job.getFinishTime())).
         _("Elapsed:", StringUtils.formatTime(
             Times.elapsed(job.getStartTime(), job.getFinishTime(), false)));
diff --git a/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-hs/src/main/java/org/apache/hadoop/mapreduce/v2/hs/webapp/HsJobsBlock.java b/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-hs/src/main/java/org/apache/hadoop/mapreduce/v2/hs/webapp/HsJobsBlock.java
index 1d6f486..10de40d 100644
--- a/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-hs/src/main/java/org/apache/hadoop/mapreduce/v2/hs/webapp/HsJobsBlock.java
+++ b/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-hs/src/main/java/org/apache/hadoop/mapreduce/v2/hs/webapp/HsJobsBlock.java
@@ -76,7 +76,7 @@
       JobInfo job = new JobInfo(j);
       jobsTableData.append("[\"")
       .append(dateFormat.format(new Date(job.getSubmitTime()))).append("\",\"")
-      .append(dateFormat.format(new Date(job.getStartTime()))).append("\",\"")
+      .append(job.getStartTimeStr()).append("\",\"")
       .append(dateFormat.format(new Date(job.getFinishTime()))).append("\",\"")
       .append("<a href='").append(url("job", job.getId())).append("'>")
       .append(job.getId()).append("</a>\",\"")
diff --git a/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-hs/src/main/java/org/apache/hadoop/mapreduce/v2/hs/webapp/dao/JobInfo.java b/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-hs/src/main/java/org/apache/hadoop/mapreduce/v2/hs/webapp/dao/JobInfo.java
index 3fbb8d2..eddbccf 100644
--- a/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-hs/src/main/java/org/apache/hadoop/mapreduce/v2/hs/webapp/dao/JobInfo.java
+++ b/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-hs/src/main/java/org/apache/hadoop/mapreduce/v2/hs/webapp/dao/JobInfo.java
@@ -18,6 +18,7 @@
 package org.apache.hadoop.mapreduce.v2.hs.webapp.dao;
 
 import java.util.ArrayList;
+import java.util.Date;
 import java.util.List;
 import java.util.Map;
 
@@ -43,6 +44,7 @@
 @XmlRootElement(name = "job")
 @XmlAccessorType(XmlAccessType.FIELD)
 public class JobInfo {
+  private static final String NA = "N/A";
 
   protected long submitTime;
   protected long startTime;
@@ -226,6 +228,16 @@ public long getStartTime() {
     return this.startTime;
   }
 
+  public String getStartTimeStr() {
+    String str = NA;
+
+    if (startTime >= 0) {
+      str = new Date(startTime).toString();
+    }
+
+    return str;
+  }
+
   public long getFinishTime() {
     return this.finishTime;
   }
diff --git a/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-hs/src/test/java/org/apache/hadoop/mapreduce/v2/hs/webapp/dao/TestJobInfo.java b/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-hs/src/test/java/org/apache/hadoop/mapreduce/v2/hs/webapp/dao/TestJobInfo.java
index b4f8d5a..e75cce7 100644
--- a/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-hs/src/test/java/org/apache/hadoop/mapreduce/v2/hs/webapp/dao/TestJobInfo.java
+++ b/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-hs/src/test/java/org/apache/hadoop/mapreduce/v2/hs/webapp/dao/TestJobInfo.java
@@ -22,6 +22,7 @@
 import static org.mockito.Mockito.when;
 
 import java.io.IOException;
+import java.util.Date;
 import java.util.HashMap;
 
 import org.junit.Assert;
@@ -135,5 +136,28 @@ public void testAverageReduceTime() {
     JobInfo jobInfo = new JobInfo(job);
   
     Assert.assertEquals(11L, jobInfo.getAvgReduceTime().longValue());
-  }  
+  }
+
+  @Test
+  public void testGetStartTimeStr() {
+    JobReport jobReport = mock(JobReport.class);
+    when(jobReport.getStartTime()).thenReturn(-1L);
+
+    Job job = mock(Job.class);
+    when(job.getReport()).thenReturn(jobReport);
+    when(job.getName()).thenReturn("TestJobInfo");
+    when(job.getState()).thenReturn(JobState.SUCCEEDED);
+
+    JobId  jobId = MRBuilderUtils.newJobId(1L, 1, 1);
+    when(job.getID()).thenReturn(jobId);
+
+    JobInfo jobInfo = new JobInfo(job);
+    Assert.assertEquals("N/A", jobInfo.getStartTimeStr());
+
+    Date date = new Date();
+    when(jobReport.getStartTime()).thenReturn(date.getTime());
+
+    jobInfo = new JobInfo(job);
+    Assert.assertEquals(date.toString(), jobInfo.getStartTimeStr());
+  }
 }
-- 
1.7.9.5

